frame/dapps-staking/src/pallet/mod.rs

```diff
@@ -0,0 +1,846 @@
+                let previous_era = Self::current_era();
+                Self::reward_balance_snapshoot(previous_era);
+                let next_era = previous_era + 1;
+                CurrentEra::<T>::put(next_era);
+                let zero_balance: BalanceOf<T> = Default::default();
+                BlockRewardAccumulator::<T>::put(zero_balance);
+                if force_new_era {
+                    ForceEra::<T>::put(Forcing::ForceNone);
+                }
+                Self::deposit_event(Event::<T>::NewDappStakingEra(next_era));
+            }
+
+            // just return the weight of the on_finalize.
+            T::DbWeight::get().reads(1)
+        }
+    }
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        /// register contract into staking targets.
+        /// contract_id should be ink! or evm contract.
+        ///
+        /// Any user can call this function.
+        /// However, caller have to have deposit amount.
+        /// TODO: weight, and add registrationFee
+        #[pallet::weight(1_000_000_000)]
+        pub fn register(
+            origin: OriginFor<T>,
+            contract_id: SmartContract<T::AccountId>,
+        ) -> DispatchResultWithPostInfo {
+            let developer = ensure_signed(origin)?;
+
+            ensure!(
+                !RegisteredDevelopers::<T>::contains_key(&developer),
+                Error::<T>::AlreadyUsedDeveloperAccount
+            );
+            ensure!(
+                !RegisteredDapps::<T>::contains_key(&contract_id),
+                Error::<T>::AlreadyRegisteredContract
+            );
+            ensure!(
+                Self::is_contract_valid(&contract_id),
+                Error::<T>::ContractIsNotValid
+            );
+
+            if Self::pre_approval_is_enabled() {
+                Self::pre_approved_contracts()
+                    .contains(&contract_id)
+                    .then(|| true)
+                    .ok_or(Error::<T>::RequiredContractPreApproval)?;
+            }
+            RegisteredDapps::<T>::insert(contract_id.clone(), developer.clone());
+            RegisteredDevelopers::<T>::insert(&developer, contract_id.clone());
+
+            Self::deposit_event(Event::<T>::NewContract(developer, contract_id));
+
+            Ok(().into())
+        }
+
+        /// add contract address to the pre-approved list.
+        /// contract_id should be ink! or evm contract.
+        ///
+        /// Sudo call is required
+        #[pallet::weight(1_000_000)]
+        pub fn contract_pre_approval(
+            origin: OriginFor<T>,
+            contract_id: SmartContract<T::AccountId>,
+        ) -> DispatchResultWithPostInfo {
+            ensure_root(origin)?;
+            ensure!(
+                !RegisteredDapps::<T>::contains_key(&contract_id),
+                Error::<T>::AlreadyRegisteredContract
+            );
+
+            let pre_approved_contracts = Self::pre_approved_contracts();
```

Comment:
I believe you could check existence of key on storage instead of fetch them all into memory.

Reply:
The `StorageMap` has a method `contains_key`, but I do not see that `StorageValue` has the `contains`.
https://crates.parity.io/frame_support/storage/trait.StorageValue.html
please advice

Reply:
I guess we need a set but I haven't seen it in Substrate (or I cannot find it).

Maybe we can use a Map that maps AccountId => ()?

Reply:
Right, StorageMap could be more optimal for large sets